src/lib.rs

```diff
@@ -1,171 +1,164 @@
+extern crate time;
 
-#![feature(asm)]
-#![feature(const_fn)]
-#![feature(core_intrinsics)]
-#![feature(integer_atomics)]
-#![feature(libc)]
+use std::net::UdpSocket;
 
-extern crate libc;
+mod pcg32;
 
-use libc::{c_void, MSG_DONTWAIT};
-use std::io::{Cursor, Write};
-use std::net::UdpSocket;
-use std::os::unix::io::{AsRawFd, IntoRawFd};
-use std::mem;
-use std::sync::atomic::{AtomicI32, Ordering};
-
-const SOCKET_FD_INIT: i32 = -1;
-
-static mut SOCKET_FD: AtomicI32 = AtomicI32::new(SOCKET_FD_INIT);
-static mut PREFIX: [u8; 576] = [0; 576];
-static mut PREFIX_LEN: usize = 0;
-
-pub fn init(address: &str, prefix: &str) -> Result<(), std::io::Error>
-{
-    let socket = UdpSocket::bind("0:0")?; // NB: CLOEXEC by default
-    socket.set_nonblocking(true)?;
-    socket.connect(address)?;
-    // so, we could protect this like log does, using a CAS on a state
-    // and so on, in the case that you choose to call `init` from a
-    // thousand different threads at once.  but, that's not a likely
-    // use-case here, so we do all the work up front, and just use the
-    // CAS to see if we won the race or not.  if we lost, free what we
-    // allocated and go home.
-    if SOCKET_FD_INIT != unsafe { SOCKET_FD.compare_and_swap(SOCKET_FD_INIT,
-                                                             socket.as_raw_fd(),
-                                                             Ordering::SeqCst) } {
-        return Ok(())
-    }
-    unsafe {
-        for (i,c) in prefix.bytes().enumerate() {
-            PREFIX[i] = c;
-        }
-        PREFIX_LEN = prefix.len();
-        if PREFIX[PREFIX_LEN-1] != b'.' {
-            PREFIX[PREFIX_LEN] = b'.';
-            PREFIX_LEN += 1;
-        }
-    }
-    socket.into_raw_fd();       // so this doesn't get closed
-    // XXX, that's technically a leaked fd; by design, basically.
-    Ok(())
-}
+const MAX_UDP_PAYLOAD: usize = 576;
 
-#[doc(hidden)]
-pub fn send(strings: &[&[u8]]) {
-    // ideally we'd construct an iov here
-    let fd = unsafe { SOCKET_FD.load(Ordering::Relaxed) };
-    if SOCKET_FD_INIT == fd { return }
-    let mut buf: [u8; 576];
-    buf = unsafe { mem::uninitialized() };
-    let len = {
-        let mut cursor = Cursor::new(&mut buf[..]);
-        let _ = cursor.write_all(unsafe { &PREFIX[..PREFIX_LEN] });
-        for s in strings { let _ = cursor.write_all(s); }
-        cursor.position() as usize
-    };
-    unsafe {
-        libc::send(fd, buf.as_ptr() as *mut c_void, len, MSG_DONTWAIT);
-    }
+pub struct StatsdClient {
```

Comment:
This was probably using a global state before for convenience reason.
For instance, being able to write a single line to send the data [like here](https://github.com/adgear/acr-service/blob/f9cf67bbc090a136d4777e0b5b16c22807dcb167/metamorphosis/src/consumer.rs#L72).
For the same reason, it might be worth it providing a singleton to be able to use is like:
```rust
statsd().increment("boring", 22, half_rate); // if using the proposed change of using methods instead of macro.
```
or even a macro.